conda_forge_tick/lazy_json_backends.py

```diff
@@ -152,13 +172,119 @@ def hkeys(self, name):
             fnames = glob.glob(os.path.join(name, "**/*.json"), recursive=True)
         return [os.path.basename(fname)[:-jlen] for fname in fnames]
 
-    def hget(self, name, key):
+    def hget(self, name: str, key: str) -> str:
         sharded_path = get_sharded_path(f"{name}/{key}.json")
         with open(sharded_path) as f:
             data_str = f.read()
         return data_str
 
 
+class GithubLazyJsonBackend(LazyJsonBackend):
+    """
+    Read-only backend that makes live requests to https://raw.githubusercontent.com URLs for serving
+    JSON files.
+
+    Any write operations are ignored!
+    """
+
+    _write_warned = False
+    _n_requests = 0
+
+    def __init__(self) -> None:
+        self.base_url = GITHUB_GRAPH_BACKEND_BASE_URL
+
+    @property
+    def base_url(self) -> str:
+        return self._base_url
+
+    @base_url.setter
+    def base_url(self, value: str) -> None:
+        if not value.endswith("/"):
+            value += "/"
+        self._base_url = value
+
+    @classmethod
+    def _ignore_write(cls) -> None:
+        if cls._write_warned:
+            return
+        logger.info(f"Note: Write operations to the GitHub online backend are ignored.")
+        cls._write_warned = True
+
+    @classmethod
+    def _inform_web_request(cls):
+        cls._n_requests += 1
+        if cls._n_requests % 20 == 0:
+            logger.info(
+                f"Made {cls._n_requests} requests to the GitHub online backend.",
+            )
+        if cls._n_requests == 20:
+            logger.warning(
+                f"Using the GitHub online backend for making a lot of requests is not recommended.",
+            )
+
+    def transaction_context(self) -> "GithubLazyJsonBackend":
+        # context not required
+        yield self
+
+    def snapshot_context(self) -> "GithubLazyJsonBackend":
+        # context not required
+        yield self
+
+    def hexists(self, name: str, key: str) -> bool:
+        self._inform_web_request()
+        url = urllib.parse.urljoin(self.base_url, f"{name}/{key}.json")
+        status = requests.head(url).status_code
+
+        if status == 200:
+            return True
+        if status == 404:
+            return False
+
+        raise RuntimeError(f"Unexpected status code {status} for {url}")
+
+    def hset(self, name: str, key: str, value: str) -> None:
+        self._ignore_write()
```

Comment:
As you are downloading the file to disk anyway, you could write the changes to your local copy. If you then keep the unchanged original file and the "dirty" file, you could then easily compare the changes the bot has done to cf graph.

Reply:
Noted!